test/FsAutoComplete.Tests.Lsp/CodeFixTests.fs

```diff
@@ -205,11 +205,170 @@ let missingInstanceMemberTests state =
     })
   ]
 
+let unusedValueTests state =
+  let server =
+    async {
+      let path = Path.Combine(__SOURCE_DIRECTORY__, "TestCases", "UnusedValue")
+      let cfg = { defaultConfigDto with UnusedDeclarationsAnalyzer = Some true }
+      let! (server, events) = serverInitialize path cfg  state
+      do! waitForWorkspaceFinishedParsing events
+      let path = Path.Combine(path, "Script.fsx")
+      let tdop : DidOpenTextDocumentParams = { TextDocument = loadDocument path }
+      do! server.TextDocumentDidOpen tdop
+      let! diagnostics =
+          events
+          |> waitForFsacDiagnosticsForFile "Script.fsx"
+          |> AsyncResult.bimap (fun _ -> failtest "Should have had errors") id
+      return (server, path, diagnostics)
+    }
+    |> Async.Cache
+
+  let (|ActRefactor|_|) title newText range (action : CodeAction) =
+    match action with
+    | { Title = title'
+        Kind = Some "refactor"
+        Edit = {
+          DocumentChanges = Some [| { Edits = [| {
+            NewText = newText'
+            Range = range'
+          } |] } |]
+        } } when title'= title && newText' = newText && range' = range
+        -> Some action
+    | _ -> None
+
+  let (|ActReplace|_|) = (|ActRefactor|_|) "Replace with _" "_"
+
+  let (|ActPrefix|_|) oldText = (|ActRefactor|_|) "Prefix with _" $"_{oldText}"
+
+  let canReplaceUnusedSelfReference = testCaseAsync "can replace unused self-reference" (async {
+    let! server, file, diagnostics = server
+    let detected = {
+        CodeActionParams.TextDocument = { Uri = Path.FilePathToUri file }
+        Range = {
+          //NOTE range is ignored for triggering -- but used in response checking
+          Start = { Line = 2; Character =  9 }
+          End   = { Line = 2; Character = 13 }
+        }
+        Context = { Diagnostics = diagnostics }
```

Comment:
So here you're passing every diagnostic found for the file, but what I think the client actually does (and what I think we do in other codefix tests) is that we filter the total set of diagnostics for the file down to the ones relevant for the range you're triggering on. See [this line](https://github.com/fsharp/FsAutoComplete/pull/795/files?file-filters%5B%5D=.fs&file-filters%5B%5D=.fsx#diff-95ae3de65e95ac64cc2bd2450abb6590783db36c49f0db4729699dae58f628b7R37) for example, but the idea being that we're validating:

* that there _was_ a diagnostic fired for the intended position, and
* that the diagnostic's fix was applied by the CodeFix action

If you do this, then I think your Expect calls get much simpler, because it should just be the one text edit and you shouldn't have to filter across the whole set of proposed changes.

Reply:
This makes perfect sense... only I am still receiving multiple code actions back. So, I've moved the range check up to the pre-condition filter (per your suggestion). But I'm still having to check an array for the correct entry in the `Expect` call. Perhaps I missed something else?

Reply:
I'll try to take a look at this this weekend to close the loop here.

Reply:
I found the gap in the mental model here, I think:

When you make the codeAction call, you get an array back because there could be many possible resolutions to a particular triggering event. For some of the triggering scenarios, you (rightfully!) would get two potential fixes back from FSAC. I collapsed two of the tests (for the unused binding and the unused parameter) to assert on the contents of the entire array of fixes (since those codefixes result in multiple potential solutions) and now everything is green (as it was before), but more compact.

Reply:
Thanks! That's completely reasonable (and looks a bit tidier, too). 👍